client/network/src/discovery.rs

```diff
@@ -516,127 +518,84 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 		list
 	}
 
-	fn inject_address_change(
-		&mut self,
-		peer_id: &PeerId,
-		connection_id: &ConnectionId,
-		old: &ConnectedPoint,
-		new: &ConnectedPoint,
-	) {
-		self.kademlia.inject_address_change(peer_id, connection_id, old, new)
-	}
-
-	fn inject_connection_established(
-		&mut self,
-		peer_id: &PeerId,
-		conn: &ConnectionId,
-		endpoint: &ConnectedPoint,
-		failed_addresses: Option<&Vec<Multiaddr>>,
-		other_established: usize,
-	) {
-		self.num_connections += 1;
-		self.kademlia.inject_connection_established(
-			peer_id,
-			conn,
-			endpoint,
-			failed_addresses,
-			other_established,
-		)
-	}
+	fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+		match event {
+			FromSwarm::ConnectionEstablished(e) => {
+				self.num_connections += 1;
+				self.kademlia.on_swarm_event(FromSwarm::ConnectionEstablished(e));
```

Comment:
A little bit of bikeshedding: you can use @ to re-bind FromSwarm::ConnectionEstablished(e) in the match arm.

Reply:
this doesn't work because 

```
error[E0308]: mismatched types
   --> client/network/src/request_responses.rs:376:42
    |
376 |                     NetworkBehaviour::on_swarm_event(p, e);
    |                     --------------------------------    ^ expected struct `RequestResponseHandler`, found struct `MultiHandler`
    |                     |
    |                     arguments to this function are incorrect
    |
    = note: expected enum `FromSwarm<'_, RequestResponseHandler<_>>`
               found enum `FromSwarm<'_, MultiHandler<std::string::String, RequestResponseHandler<_>>>`
```